src/NServiceBus.Core/Routing/DistributionContext.cs

```diff
@@ -0,0 +1,49 @@
+namespace NServiceBus.Routing
+{
+    using System.Collections.Generic;
+    using Extensibility;
+    using Pipeline;
+
+    /// <summary>
+    /// The distribution context for the DistributionStrategy implementers.
+    /// </summary>
+    public class DistributionContext
+    {
+        /// <summary>
+        /// Creates a new distribution context.
+        /// </summary>
+        public DistributionContext(string[] receiverAddresses, OutgoingLogicalMessage message, string messageId, Dictionary<string, string> headers, ContextBag context)
+        {
+            ReceiverAddresses = receiverAddresses;
+            Message = message;
+            MessageId = messageId;
+            Headers = headers;
+            Context = context;
+        }
+
+        /// <summary>
+        /// The receiver addresses that can be taken into account for distribution.
+        /// </summary>
+        public string[] ReceiverAddresses { get; }
+
+        /// <summary>
+        /// The id of the outgoing message.
+        /// </summary>
+        public string MessageId { get; private set; }
+
+        /// <summary>
+        /// The headers of the outgoing message.
+        /// </summary>
+        public Dictionary<string, string> Headers { get; }
```

Comment:
hm it's a different thing on the errorcontext? It gives access to the incoming messages headers. Modifying those doesn't change anything since those would be reverted anyway. It's different here as it would allow you to modify headers of an outgoing message where you shouldn't do that (or is that a valid use case?)

for the high-performance, we could make the readonly dictionary field lazy-initialized and only allocate when the headers are actually accessed?